internal/caddy/ingress/redirect_test.go

```diff
@@ -0,0 +1,113 @@
+			expectedConfigPath: "test_data/redirect_default.json",
+			annotations: map[string]string{
+				"caddy.ingress.kubernetes.io/permanent-redirect": "http://example.com",
+			},
+		},
+		{
+			name:               "Check permanent redirect with 'permanent' redirect code",
+			expectedConfigPath: "test_data/redirect_permanent.json",
+			annotations: map[string]string{
+				"caddy.ingress.kubernetes.io/permanent-redirect":      "http://example.com",
+				"caddy.ingress.kubernetes.io/permanent-redirect-code": "permanent",
+			},
+		},
+		{
+			name:               "Check permanent redirect with 'temporary' redirect code",
+			expectedConfigPath: "test_data/redirect_temporary.json",
+			annotations: map[string]string{
+				"caddy.ingress.kubernetes.io/permanent-redirect":      "http://example.com",
+				"caddy.ingress.kubernetes.io/permanent-redirect-code": "temporary",
+			},
+		},
+		{
+			name:               "Check permanent redirect with custom redirect code",
+			expectedConfigPath: "test_data/redirect_custom_code.json",
+			annotations: map[string]string{
+				"caddy.ingress.kubernetes.io/permanent-redirect":      "http://example.com",
+				"caddy.ingress.kubernetes.io/permanent-redirect-code": "308",
+			},
+		},
+		{
+			name:               "Check permanent redirect with invalid custom redirect code",
+			expectedConfigPath: "",
+			annotations: map[string]string{
+				"caddy.ingress.kubernetes.io/permanent-redirect":      "http://example.com",
+				"caddy.ingress.kubernetes.io/permanent-redirect-code": "502",
+			},
+			expectedError: "redir code not in the 3xx range or 401: '502'",
+		},
+		{
+			name:               "Check permanent redirect with invalid custom redirect code string",
+			expectedConfigPath: "",
+			annotations: map[string]string{
+				"caddy.ingress.kubernetes.io/permanent-redirect":      "http://example.com",
+				"caddy.ingress.kubernetes.io/permanent-redirect-code": "randomstring",
+			},
+			expectedError: "not a supported redir code type or not valid integer: 'randomstring'",
+		},
+		{
+			name:               "Check permanent redirect with 401 as redirect code",
+			expectedConfigPath: "test_data/redirect_401.json",
+			annotations: map[string]string{
+				"caddy.ingress.kubernetes.io/permanent-redirect":      "http://example.com",
+				"caddy.ingress.kubernetes.io/permanent-redirect-code": "401",
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			input := converter.IngressMiddlewareInput{
+				Ingress: &networkingv1.Ingress{
+					ObjectMeta: metav1.ObjectMeta{
+						Annotations: test.annotations,
+					},
+				},
+				Route: &caddyhttp.Route{},
+			}
+
+			route, err := rp.IngressHandler(input)
+			if test.expectedError != "" {
+				if assert.Error(t, err) {
+					assert.EqualError(t, err, test.expectedError)
+				}
+			} else {
+				assert.NoError(t, err, "unable to generate the route.")
+			}
+
+			if test.expectedError == "" {
```

Comment:
I'm not confident with this way to check if we want the test to fail or not (having the `expectedError` field). If the field is present, we check if the error returned from the ingress plugin is what we expect.

If the field is not present, then we proceed with comparing the generated config and the sample config file if they match.

Reply:
Can I suggest to split this into two tests? One for the entry that do not expect any error and one for the group that expect errors? This think will make the tests more understandable and more clean in term of implementation, something like:
```
func TestRedirectConvertToCaddyConfig(t *testing.T) {
    // ...
}

func TestMisconfiguredRedirectConvertToCaddyConfig(t *testing.T) {
    // ...
}
```